data/bundled.lean

```diff
@@ -0,0 +1,53 @@
+/-
+Copyright (c) 2018 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison, Johannes Hölzl, Reid Barton, Sean Leather
+
+Bundled type and type class instance.
+-/
+
+universes u v
+
+/-- `bundled` is a type bundled with a type class instance for that type. Only
+the type class is exposed as a parameter. -/
+structure bundled (c : Sort u → Sort v) : Sort (max (u+1) v) :=
```

Comment:
Should we parameterise `Sort u`? After all `Sort u` is a `Type u`. If we change this to `bundled {X : Type u} (c : X → Sort v) := ...` then this would generalise the previous version of `bundled` an also `subtype`. Is that right? Would that be useful?

Reply:
Indeed, you're right. The change is simple. I'm not sure if it's useful.

```diff
@@ -10,8 +10,8 @@ universes u v
  /-- `bundled` is a type bundled with a type class instance for that type. Only
 the type class is exposed as a parameter. -/
-structure bundled (c : Sort u → Sort v) : Sort (max (u+1) v) :=
-mk' :: (α : Sort u) (inst : c α)
+structure bundled {τ : Type u} (c : τ → Sort v) : Sort (max (u+1) v) :=
+mk' :: (α : τ) (inst : c α)
 
 /-
 Note on the definition of `bundled`:
@@ -31,14 +31,14 @@ below.
 -/
 
 namespace bundled
-variables {c d : Sort u → Sort v} {α : Sort u}
+variables {τ : Type u} {c d : τ → Sort v} {α : Sort u}
 
 /-- A convenient wrapper around `bundled.mk'` with an implicit instance. -/
-def mk (α : Sort u) [h : c α] : bundled c :=
+def mk (α : τ) [h : c α] : bundled c :=
 mk' α h
 
 instance : has_coe_to_sort (bundled c) :=
-{ S := Sort u, coe := bundled.α }
+{ S := τ, coe := bundled.α }
 
 /-- Map over the bundled instance -/
 def map (f : ∀ {α}, c α → d α) (b : bundled c) : bundled d :=
```

Now that I'm thinking about it, why doesn't `sigma` or `psigma` work in place of `bundled`?

Reply:
I think the point is just that we want a special name to key on, when providing helpers to construct categories of bundled things.

Reply:
Nothing “goes wrong.” I suppose it's just a matter of taste. See #407 for the use-`sigma`-instead option.

Reply:
One thing I don't like about using `sigma` this way is that I think it's very natural to provide an instance of category for both subtype and for sigma, that ignores the condition/data --- this is a good way to think about full subcategories. (In practice, when mathematicians talk about full subcategories, they are sometimes actually talking about things based on sigma, rather than subtype.)